### PR TITLE
Cleanup serial handling and logs

### PR DESCRIPTION
**Pull request by GMagician**
Should address #84 and #85

### CODE REVIEW

File: ESP8266WirelessPrintAsync/ESP8266WirelessPrintAsync.ino [hunk 1]
```diff
@@ -825,59 +851,55 @@ void ReceiveResponses() {
...
+        lastReceivedResponse = serialResponse;
+        lineStartPos = 0;
+        serialResponse = "";
       }
     }
   }
 
   if (!commandQueue.isAckEmpty() && (signed)(serialReceiveTimeoutTimer - millis()) <= 0) {  // Command has been lost by printer, buffer has been freed
-    telnetSend("< #TIMEOUT#");
+    telnetSend("<#TIMEOUT#");
```
Comment (probonopd) [inline ESP8266WirelessPrintAsync/ESP8266WirelessPrintAsync.ino]: Let's use `<` only for strings received from the printer. As `#TIMEOUT#` is not a string received from the printer, it should not start with `<`

File: ESP8266WirelessPrintAsync/ESP8266WirelessPrintAsync.ino [hunk 2]
```diff
@@ -825,59 +851,55 @@ void ReceiveResponses() {
...
-        lineStartPos = serialResponse.length();
-        telnetSend("Unhandled line received:");
-        telnetSend("<" + String(serialResponse) );
+        restartSerialTimeout();     // Something has been received so clear timeout (printer and communication are live
+
+        if (parseTemperatures(serialResponse))
+          responseDetail = "autotemp";
+        else if (parsePosition(serialResponse))
+          responseDetail = "position";
+        else if (serialResponse.startsWith("echo:busy"))
```
Comment (probonopd) [inline ESP8266WirelessPrintAsync/ESP8266WirelessPrintAsync.ino]: No space after the ":" here? Compare to the next line

Comment (GMagician) [reply ESP8266WirelessPrintAsync/ESP8266WirelessPrintAsync.ino]: in marlin message is "echo:busy: xxxxxx". only space after busy ':'
OT what do you do to send these post? with code reference? help...

**Comment by probonopd** [comment]
Hmm, https://transfer.sh/ is having issues at the moment, so the build artifacts of PRs don't get uploaded at the moment. Can you enable https://github.com/probonopd/uploadtool in your repo by setting `GITHUB_TOKEN` on https://travis-ci.com/GMagician/WirelessPrinting/settings as described in https://github.com/probonopd/uploadtool?

**Comment by GMagician** [comment]
that should be done

**Comment by probonopd** [comment]
This does not seem to work at all for me; getting

```
Trying 192.168.0.27...
Connected to 192.168.0.27.
Escape character is '^]'.
<#TIMEOUT#
<#TIMEOUT#
<#TIMEOUT#
<#TIMEOUT#
<#TIMEOUT#
<#TIMEOUT#
```

right from the beginning.

File: ESP8266WirelessPrintAsync/ESP8266WirelessPrintAsync.ino (hunk 2 above)
Comment (GMagician) [reply ESP8266WirelessPrintAsync/ESP8266WirelessPrintAsync.ino]: Both line are correct, checked Marlin 2, cold has space busy not

**Comment by probonopd** [comment]
Here is a full log from 1.x for comparison. Things to note:

* Every line coming from the printer is displayed, prefixed by `< ` Regardless of whether they could be interpreted or not
* Every line sent to the printer is displayed, prefixed by `> `
* The temperatures which the printer sends were understood, see below. No special handling of the Prusa firmware was needed
* We did not send anything to the printer before the printer responded with "ok"
* There were no timeouts

```
{
  "temperature": {
    "tool0": {
      "actual": 218.39 E:0 B:57.5,
      "target": :218.39,
      "offset": 0
    },
    "bed": {
      "actual": 57.5,
      "target": :218.39,
      "offset": 0
    }
  },
  "sd": {
    "ready": 1
  },
  "state": {
    "text": "Operational",
    "flags": {
      "operational": true,
      "paused": false,
      "printing": 1,
      "sdReady": 1,
      "error": false,
      "ready": true,
      "closedOrError": false
    }
  }
}
```

**Comment by GMagician** [comment]
I'm talking about what prusa does when it is heating.
When prusa MK2 is heating respond with T:nn E:n B:nn (without ok and without busy)
in same situation marlin respond with usual M105 (T:nn/nn B:nn/nn) message (without ok but with busy)

**Comment by probonopd** [comment]
For the "still alive" timer it should not matter whether we can parse a response from the printer or not.
For the question whether we can send the next line to the printer all that matters if we got `ok` from the printer.

**Comment by probonopd** [comment]
Still getting

```
Trying 192.168.0.27...
Connected to 192.168.0.27.
Escape character is '^]'.
#TIMEOUT#
#TIMEOUT#
#TIMEOUT#
#TIMEOUT#
#TIMEOUT#
```
right from the beginning in c970472.

**Comment by GMagician** [comment]
yes...that's is baud detection that never receive ok and never go on....I'll try a different approach

**Comment by GMagician** [comment]
try nr #1

**Comment by probonopd** [comment]
This fixes baudrate detection:

```
Trying 192.168.0.27...
Connected to 192.168.0.27.
Escape character is '^]'.
Connecting at 500000
>���115
Connecting at 250000
>���115
Connecting at 115200
>���115
<FIRMWARE_NAME:Marlin V1.0.2; Sprinter/grbl mashup for gen6 FIRMWARE_URL:https://github.com/prusa3d/Prusa-i3-Plus/ PROTOCOL_VERSION:1.0 MACHINE_TYPE:Prusa i3 MK2 EXTRUDER_COUNT:1 UUID:00000000-0000-0000-0000-000000000000
 #wait more#
<FIRMWARE_NAME:Marlin V1.0.2; Sprinter/grbl mashup for gen6 FIRMWARE_URL:https://github.com/prusa3d/Prusa-i3-Plus/ PROTOCOL_VERSION:1.0 MACHINE_TYPE:Prusa i3 MK2 EXTRUDER_COUNT:1 UUID:00000000-0000-0000-0000-000000000000
ok
 ##
Connected
>M117 192.168.0.27 SD
<ok
 ##
>M300 S500 P50
<ok
 ##
>M105
<ok T:29.0 /0.0 B:28.6 /0.0 T0:29.0 /0.0 @:0 B@:0
 ##
>M105
<ok T:29.0 /0.0 B:28.6 /0.0 T0:29.0 /0.0 @:0 B@:0
 ##
```

Now trying to print something.

**Comment by probonopd** [comment]
It is now not sending stuff to the printer before we get an OK, but the `/api/printer` does not get updated with the temperature values during heating. During the whole heating phase it shows:

```
{
  "state": {
    "text": "Printing",
    "flags": {
      "operational": true,
      "paused": false,
      "printing": true,
      "pausing": false,
      "cancelling": false,
      "sdReady": false,
      "error": false,
      "ready": true,
      "closedOrError": false
    }
  },
  "temperature": {
    "tool0": {
      "actual": 29.0,
      "target": 0.0,
      "offset": 0
    },
    "bed": {
      "actual": 28.6,
      "target": 0.0,
      "offset": 0
    }
  },
  "sd": {
    "ready": false
  }
}
```

Still getting `#TIMEOUT#` when the printer starts its autocalibration.

But printing works. Log attached.

[log.txt.zip](https://github.com/probonopd/WirelessPrinting/files/2922024/log.txt.zip)

**Comment by probonopd** [comment]
Why are there many lines like

```
T:158.23 E:0 B:39.0
T:160.15 E:0 B:39.2
T:162.44 E:0 B:39.4
T:164.49 E:0 B:39.7
T:166.51 E:0 B:40.0
```

without being prefixed by `<`? Everything coming from the printer should be printed to telnet, prefixed by a `< ` as the first thing (before we even try to interpret it).

**Comment by GMagician** [comment]
> without being prefixed by <? Everything coming from the printer should be printed to telnet, prefixed by a < as the first thing (before we even try to interpret it).

that is what it should do...

**Comment by GMagician** [comment]
I think your firmware works a quite differently by marlin.. have you ever seen a T:xxx/ yy message?

**Comment by probonopd** [comment]
Instead of

> Add missing \n when message not complete 

why not start processing a line only after `\n` has been received? In the 1,x branch I did:

https://github.com/probonopd/WirelessPrinting/blob/a1dbae89cd9782f1dd7dced60d97f24e029a350d/ESP8266WirelessPrintAsync/ESP8266WirelessPrintAsync.ino#L188-L192

This has proven to be very robust for me.

> have you ever seen a T:xxx/ yy message?

See the log from the 1.x branch which I had posted in https://github.com/probonopd/WirelessPrinting/pull/87#issuecomment-468908626.

**Comment by GMagician** [comment]
> why not start processing a line only after \n has been received? In the 1.x branch I did:

Because this way I stop everything... I want to process telnet and send more than one command when serial buffer let me do this (like repetier already does)

**Comment by GMagician** [comment]
Now I think you only have some timeouts during print..right?

**Comment by probonopd** [comment]
> send more than one command when serial buffer let me do this (like repetier already does)

Isn't this adding risk of bugs? What is the advantage of sending more than one line at a time?

**Comment by probonopd** [comment]
```
Trying 192.168.0.27...
Connected to 192.168.0.27.
Escape character is '^]'.
Connecting at 500000
>115
Connecting at 250000
>115
Connecting at 115200
>115
< #wait more#
<ok #acknowledged#
Connected
>M117 192.168.0.27 SD
<ok #acknowledged#
>M300 S500 P50
<ok #acknowledged#
>M105
<ok T:28.5 /0.0 B:28.2 /0.0 T0:28.5 /0.0 @:0 B@:0 #acknowledged#
>M105
<ok T:28.5 /0.0 B:28.2 /0.0 T0:28.5 /0.0 @:0 B@:0 #acknowledged#
>M105
<ok T:28.5 /0.0 B:28.2 /0.0 T0:28.5 /0.0 @:0 B@:0 #acknowledged#
>M105
<ok T:28.5 /0.0 B:28.2 /0.0 T0:28.5 /0.0 @:0 B@:0 #acknowledged#
>M105
<ok T:28.5 /0.0 B:28.2 /0.0 T0:28.5 /0.0 @:0 B@:0 #acknowledged#
>M105
<ok T:28.5 /0.0 B:28.2 /0.0 T0:28.5 /0.0 @:0 B@:0 #acknowledged#
>M105
<ok T:28.5 /0.0 B:28.2 /0.0 T0:28.5 /0.0 @:0 B@:0 #acknowledged#
>M105
<ok T:28.5 /0.0 B:28.2 /0.0 T0:28.5 /0.0 @:0 B@:0 #acknowledged#
>M105
<ok T:28.5 /0.0 B:28.2 /0.0 T0:28.5 /0.0 @:0 B@:0 #acknowledged#
>M117 Receiving...
<ok #acknowledged#
>M117 Received
<ok #acknowledged#
>M300 S500 P50
<ok #acknowledged#
>M117 Printing...
<ok #acknowledged#
>M300 S500 P50
<ok #acknowledged#
>M105
<ok T:28.5 /0.0 B:28.2 /0.0 T0:28.5 /0.0 @:0 B@:0 #acknowledged#
>M201 X9000 Y9000 Z500 E10000 
<ok #acknowledged#
>M203 X500 Y500 Z12 E120 
<ok #acknowledged#
>M204 P2000 R1500 T2000 
<ok #acknowledged#
>M205 X10.00 Y10.00 Z0.20 E2.50 
<ok #acknowledged#
>M205 S0 T0 
<ok #acknowledged#
>M107
<ok #acknowledged#
>M115 U3.1.0 
<ok #acknowledged#
>M83  
<ok #acknowledged#
>M204 S2000 T1500 
<ok #acknowledged#
>M104 S215 
<ok #acknowledged#
>M140 S60 
<ok #acknowledged#
>M190 S60 
<T:28.50 E:0 B:28.6 #autotemp#
<T:28.50 E:0 B:28.6 #autotemp#
<T:29.00 E:0 B:28.6 #autotemp#
(...)
```

`/api/printer` is showing the correct values during heating :+1: 

When heating done and bed calibration starts: 

```
<T:212.50 E:0 B:59.7 #autotemp#
<T:212.50 E:0 B:59.7 #autotemp#
<T:212.97 E:0 B:60.0 #autotemp#
<ok #acknowledged#
>M109 S215 
<T:213.5 E:0 W:? #autotemp#
<T:214.6 E:0 W:2 #autotemp#
<T:215.8 E:0 W:1 #autotemp#
<T:216.7 E:0 W:0 #autotemp#
<ok #acknowledged#
>G28 W 
#TIMEOUT#
#TIMEOUT#
#TIMEOUT#
#TIMEOUT#
#TIMEOUT#
<ok #acknowledged#
>G80 
#TIMEOUT#
#TIMEOUT#
#TIMEOUT#
```
But the __print starts__.

[log.txt.zip](https://github.com/probonopd/WirelessPrinting/files/2922100/log.txt.zip)

**Comment by probonopd** [comment]
What do these

```
#TIMEOUT#
#TIMEOUT#
#TIMEOUT#
#TIMEOUT#
#TIMEOUT#
```
messages actually mean, other than the fact that the printer sent nothing in the last 2.5 seconds?

**Comment by probonopd** [comment]
So this solves #72 for me. Ready for merge?

**Comment by GMagician** [comment]
> messages actually mean, other than the fact that the printer sent nothing in the last 2.5 seconds?

yes.. prusa is not answering....Marlin say busy. Marli is more friendly from this point of view...

Not merge now..

**Comment by GMagician** [comment]
1) I'm wondering... #acknowledge# is too much for me what do you think?
2) on wait more it should print last received line but it doesnt...I need to fix

**Comment by probonopd** [comment]
Let's not make this too Marlin specific.

**Comment by probonopd** [comment]
> yes.. prusa is not answering....Marlin say busy. Marli is more friendly from this point of view...

If `#TIMEOUT#` does not do anything then why do we need it at all?

**Comment by probonopd** [comment]
> I'm wondering... #acknowledged# is too much for me what do you think?

It depends on what you mean by `#acknowledged#`. The fact that we have recognized that we got `ok` back? Then let's call it `#ok#`, maybe.

**Comment by GMagician** [comment]
> Then let's call it #ok#, maybe.

it's an ok message got...nut you get ok in frame and #ok# at and... why simply not say anything but the response

**Comment by probonopd** [comment]
> why simply not say anything but the response

Saying the response on telnet means "we have received this from the printer". It does not necessarily mean that we have understood the printer having said "ok". So I am for `#ok#`.

**Comment by GMagician** [comment]
```c++
  server.on("/api/version", HTTP_GET, [](AsyncWebServerRequest * request) {
    // http://docs.octoprint.org/en/master/api/version.html
    request->send(200, "application/json", "{\r\n"
                                           "  \"api\": \"" API_VERSION "\",\r\n"
                                           "  \"server\": \"" VERSION "\"\r\n"
                                           "}");  });
```

does it still should report VERSION or SKETCH_VERSION?

**Comment by probonopd** [comment]
Also, I would put all `#tags#` on separate lines, below the line starting with `>` to make clear that this is not part of what the printer has sent.

**Comment by probonopd** [comment]
> does it still should report VERSION or SKETCH_VERSION?

It reports

```
{
  "api": "0.1",
  "server": "1.3.10"
}
```

**Comment by GMagician** [comment]
I used # just to say that... if you dont enclose your response between something known you dont know if trailing spaces or not

**Comment by probonopd** [comment]
OK, I see

**Comment by GMagician** [comment]
Not sure if `substring(oldLineStartPos, newLineStartPos)` or `substring(oldLineStartPos, newLineStartPos-1)`

**Comment by probonopd** [comment]
Are you still doing all this without access to a real printer? Amazing!

**Comment by GMagician** [comment]
> Are you still doing all this without access to a real printer? Amazing!

sigh yes...my usb plug doesn't arrive. I think it as been lost between path

**Comment by probonopd** [comment]
Not good:

```
>M107
<#ok#
>M115 U3.1.0 
<#ok#
>M83  
<#ok#
>M204 S2000 T1500 
<#ok#
>M104 S215 
<#ok#
>M140 S60 
<#ok#
>M190 S60 
<#autotemp#
<#autotemp#
<#autotemp#
<#autotemp#
<#autotemp#
```

**Comment by probonopd** [comment]
> usb plug

You mean a SD shield for the WeMos D1 Mini?

**Comment by GMagician** [comment]
I used mega that needs 5V signals but mini has 3.3. I ordered a usb to usb micro converter to let me able to connect both without soldering

**Comment by probonopd** [comment]
Not exactly sure what you are trying to do but my printers deliver 5V as the input to the D1 Mini. The D1 Mini contains a voltage regulator that converts it down to the 3.3V needed by the ESP8266.

**Comment by GMagician** [comment]
I want to connect mini di mega usb. Mega has a Type C connector (I think it its name) and mini has a micro usb connector.
I have a standard cable to connect PC to mega but PC has standard usb

**Comment by GMagician** [comment]
Please another log after installation...

**Comment by probonopd** [comment]
Looking good to me now :+1: 
[log.txt.zip](https://github.com/probonopd/WirelessPrinting/files/2922136/log.txt.zip)

**Comment by GMagician** [comment]
what is needed for G80? it seems a long operation... with no feedback...

**Comment by GMagician** [comment]
> <T:212.6 E:0 W:?#autotemp#

that's wrong...is should not be recognized as autotemp....what is W:?

**Comment by probonopd** [comment]
Bed calibration

https://www.youtube.com/watch?v=rYrLT5G-a9I

**Comment by probonopd** [comment]
> what is W:?

Good question. Searching... (but again, we should not care too much about particular firmware specialties)

**Comment by GMagician** [comment]
> Bed calibration

ok I know...but no busy nor other message during it...

**Comment by GMagician** [comment]
I think you can merge it

**Comment by probonopd** [comment]
No clue what `W:` means.

**Comment by probonopd** [comment]
c73b838

[log.txt.zip](https://github.com/probonopd/WirelessPrinting/files/2922165/log.txt.zip)

**Comment by probonopd** [comment]
...and a big "thank you" @GMagician - your name is well deserved, you are a magician ;-)

**Pull request merged**